Iteration 1.

Okenamay opened the pull request:
Первый вариант. Сдаю с опозданием, с куратором согласовано.

В коде main.go внёс небольшие изменения, комментариями отметил, прошу глянуть. В коде main.go линтер ругается на []string в мапе со списком кафе. Прошу пояснить, почему в учебном материале это указание есть, а линтер говорит, что оно не нужно.

Okenamay committed:
Iteration 1.

Первый вариант. Сдаю с опозданием, с куратором согласовано.

В коде main.go внёс небольшие изменения, комментариями отметил, прошу глянуть.
В коде main.go линтер ругается на []string в мапе со списком кафе.
Прошу пояснить, почему в учебном материале это указание есть, а линтер говорит, что оно не нужно.

Okenamay commented:
Запутался, почему я сделал форк, а пулл реквест в гите Я.Практикума оказался. Видимо, надо было сделать с нуля, создать пустой репозиторий, клонировать его, и туда уже писать код.

just-musha commented:
@Okenamay супер. В идеале - ПР нужно в свой репозиторий, вместо Яндекс практикум. Там можно выбрать при создании ПРа.
 Учти при сдаче следующих работ, сейчас не буду возвращать.

just-musha commented:
> В коде main.go внёс небольшие изменения, комментариями отметил, прошу глянуть. В коде main.go линтер ругается на []string в мапе со списком кафе. Прошу пояснить, почему в учебном материале это указание есть, а линтер говорит, что оно не нужно.

На этот вопрос ответила в прошлой итерации, в системе ревью. Все ли понятно?

just-musha commented on main_test.go:
```diff
@@ -0,0 +1,62 @@
+	warningCityNotSpecified := `city not specified`                             // Сообщение о том, что город не указан.
+	warningWrongCityValue := `wrong city value`                                 // Сообщение о неверном городе.
+	assert.Equal(t, expected, responseRecorder.Code)                            // Проверяем, что статус 200 корректный.
+	assert.NotEmpty(t, responseRecorder.Body.String())                          // Проверяем, что тело ответа не пустое.
```
А это множество проверок `NotEqual` для чего тебе нужны? 

Можно ли обойтись одной проверкой - сравнить с ожидаемым значением?

Okenamay replied:
В принципе, идея была проверять при статусе ОК, что сервис не только отдаёт 200, но и что отдаёт непустой ответ, при этом не являющийся ответом о той или иной ошибке.

Конечно, формально ответ об ошибке у нас привязан к статусу 400, и такая проверка не нужна, но я зачем-то сделал её.

Okenamay replied:
Точно, можно же сравнить с выдачей на ?count=2&city=moscow

Okenamay replied:
У нас фиксированный запрос, и мы ожидаем на него детерменированный ответ.

Okenamay replied:
Исправил.

just-musha commented on main_test.go:
```diff
@@ -0,0 +1,62 @@
+
+	expected := `wrong city value`                                // Ожидаемое сообщение.
+	assert.Equal(t, responseRecorder.Code, http.StatusBadRequest) // Проверяем, что статус 400 корректный.
+	assert.Equal(t, expected, responseRecorder.Body.String())     // Проверяем, что сообщение правильное.
```
:+1: Супер, тест проверяет все что нужно

Okenamay replied:
Это радует!

just-musha commented on main_test.go:
```diff
@@ -0,0 +1,62 @@
+	expected := http.StatusOK                                     // Ожидаемое сообщение о статусе.
+	expectedList := strings.Join(totalList[:totalCount], ", ")    // Ожидаемый ответ.
+	assert.Equal(t, expected, responseRecorder.Code)              // Проверяем, что статус 200 корректный.
+	assert.NotEmpty(t, responseRecorder.Body.String())            // Проверяем, что тело ответа не пустое.
```
Здесь аналогичный вопрос про NotEqual, можно ли упростить тест?

Okenamay replied:
Аналогично, запрос фиксированный, ожидаемый ответ детерминированный.

Okenamay replied:
Поправил.

just-musha reviewed (changes requested):
Привет! Хорошее решение - есть нужные проверки, но предлагаю подумать - можно ли чуть оптимальнее сделать и упростить код. Удачи!

Okenamay committed:
Iteration 2.

Упростил первый и третий тесты.
Убрал избыточные проверки.
(Если сравниваем ответ с шаблоном, то проверка на пустой ответ ненужна и т.п.)

Okenamay commented:
Кажется, исправил всё, и закоммитил куда надо. Да уж, честное слово, намудрил в самом начале с форком и веткой – теперь вот оно как.

just-musha commented on main_test.go:
```diff
@@ -43,7 +36,7 @@ func TestMainHandlerWhenWrongCityValue(t *testing.T) {
 	totalCount := 4
-	totalList := []string{"Мир кофе", "Сладкоежка", "Кофе и завтраки", "Сытый студент"}
+	expectedList := `Мир кофе, Сладкоежка, Кофе и завтраки, Сытый студент` // Ожидаемый ответ.
 	req := httptest.NewRequest("GET", "/cafe?count=20&city=moscow", nil)
```
Можешь ещё посмотреть как добавлять параметры через query.Encode() - более читаемо и безопасно (с точки зрения если есть символы недопустимые в url, например в случае города это мб пробел в названии) https://www.urlencoder.io/golang/

Okenamay replied:
Спасибо, почитал, утром перечитаю на свежую голову.

just-musha reviewed (approved):
Отлично, получилось упростить, тесты легкие и читаемые. Принято!
